sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnosticsContext.java

```diff
@@ -0,0 +1,649 @@
+        synchronized (this.spanName) {
+            boolean hasCompletedOperation = this.isCompleted.compareAndSet(false, true);
+            if (hasCompletedOperation) {
+                this.recordOperation(statusCode, subStatusCode, actualItemCount, finalError);
+            }
+
+            return hasCompletedOperation;
+        }
+    }
+
+    synchronized void recordOperation(int statusCode, int subStatusCode, Integer actualItemCount, Throwable finalError) {
+        synchronized (this.spanName) {
+            this.statusCode = statusCode;
+            this.subStatusCode = subStatusCode;
+            this.finalError = finalError;
+            if (actualItemCount != null) {
+                if (!this.actualItemCount.compareAndSet(-1, actualItemCount)) {
+                    this.actualItemCount.addAndGet(actualItemCount);
+                }
+            }
+            this.duration = Duration.between(this.startTime, Instant.now());
+            this.cachedRequestDiagnostics = null;
+        }
+    }
+
+    String getRequestDiagnostics() {
+        ObjectNode ctxNode = mapper.createObjectNode();
+
+        ctxNode.put("spanName", this.spanName);
+        ctxNode.put("account", this.accountName);
+        ctxNode.put("db", this.databaseName);
+        if (!this.collectionName.isEmpty()) {
+            ctxNode.put("container", this.collectionName);
+        }
+        ctxNode.put("resource", this.resourceType.toString());
+        ctxNode.put("operation", this.operationType.toString());
+        if (!this.operationId.isEmpty()) {
+            ctxNode.put("operationId", this.operationId);
+        }
+        ctxNode.put("consistency", this.consistencyLevel.toString());
+        ctxNode.put("status", this.statusCode);
+        if (this.subStatusCode != 0) {
+            ctxNode.put("subStatus", this.subStatusCode);
+        }
+        ctxNode.put("RUs", this.totalRequestCharge);
+        ctxNode.put("maxRequestSizeInBytes", this.maxRequestSize);
+        ctxNode.put("maxResponseSizeInBytes", this.maxResponseSize);
+
+        if (this.maxItemCount != null) {
+            ctxNode.put("maxItems", this.maxItemCount);
+        }
+
+        if (this.actualItemCount.get() >= 0) {
+            ctxNode.put("actualItems", this.actualItemCount.get());
+        }
+
+        if (this.finalError != null) {
+            ctxNode.put("exception", this.finalError.toString());
+        }
+
+        if (this.diagnostics != null && this.diagnostics.size() > 0) {
+            ArrayNode diagnosticsNode = ctxNode.putArray("diagnostics");
+            for (CosmosDiagnostics d: this.diagnostics) {
+
+                ObjectNode childNode = mapper.createObjectNode();
+                d.fillCosmosDiagnostics(childNode, null);
+
+                diagnosticsNode.add(childNode);
+            }
+        }
+
+        try {
+            return mapper.writeValueAsString(ctxNode);
+        } catch (JsonProcessingException e) {
+            return "{ \"exception\": \"" + e + "\" }";
```

Comment:
probably need to json-escape e.toString()

Reply:
Done